app/queries/control_rate_query_v2.rb

```diff
@@ -0,0 +1,27 @@
+class ControlRateQueryV2
+  def controlled(region)
+    ReportingPipeline::PatientStatesPerMonth
+      .where(hypertension: "yes", htn_care_state: "under_care")
+      .where("assigned_#{region.region_type}_region_id" => region.id)
+      .where("months_since_registration >= ?", Reports::REGISTRATION_BUFFER_IN_MONTHS)
+      .where(htn_treatment_outcome_in_last_3_months: :controlled)
+  end
+
+  def controlled_counts(region, range: nil)
+    time_range = (range.begin.start_time..range.end.end_time)
+    controlled(region).group_by_period(:month, :month_date, range: time_range, format: Period.formatter(:month)).count
+  end
+
+  def uncontrolled(region)
+    ReportingPipeline::PatientStatesPerMonth
+      .where("assigned_#{region.region_type}_region_id" => region.id)
+      .where(hypertension: "yes", htn_care_state: "under_care")
+      .where("months_since_registration >= ?", Reports::REGISTRATION_BUFFER_IN_MONTHS)
+      .where(htn_treatment_outcome_in_last_3_months: :uncontrolled)
+  end
```

Comment:
I'm afraid of writing out the controlled and uncontrolled queries (then the missed-visits and visited-no-bp queries) as independent functions. There is an inherent coupling between them, and we should strive to maintain that.

For example, if the denominator of the query changes for one treatment outcome, it is important to change it in other places as well. Any divergence in these queries will likely introduce data variance on the dashboard which we are actively trying to avoid.

How do you feel about having a single "treatment outcome" query, that we can read any treatment outcome number from? I had spiked out one approach in the other PR: https://github.com/simpledotorg/simple-server/pull/2619/files#diff-88aaebae700a1b0c64581250b0d554092bd7e8adbc64a85fc427a0fad71a7024R88-R112

Reply:
Not sure I follow, this doesn't involved the denominator at all.  This is just the numerator.  

All the queries you mention use the same denominator currently in repository, and will continue to with the new matviews as well.

Reply:
For reference, here is the denominator used for calculating rates - https://github.com/simpledotorg/simple-server/blob/04ece50ae87a8885aa723e3b765e18ca4103ac15/app/models/reports/repository.rb#L278

Reply:
Trying to illustrate what I mean by keeping the common/denominator in one place:
```ruby
  def common(maybe_region)
    region = maybe_region.region
    ReportingPipeline::PatientStatesPerMonth
      .where(hypertension: "yes", htn_care_state: "under_care")
      .where("assigned_#{region.region_type}_region_id" => region.id)
      .where("months_since_registration >= ?", Reports::REGISTRATION_BUFFER_IN_MONTHS)
  end
  
  def controlled(maybe_region)
    common(maybe_region).where(htn_treatment_outcome_in_last_3_months: :controlled)
  end

  def uncontrolled(maybe_region)
    common(maybe_region).where(htn_treatment_outcome_in_last_3_months: :uncontrolled)
  end
```

We should also be using the same query for the denominator in treatment outcome rates as well.

Reply:
Sure, I can refactor towards something like that.

`group_date` gives us consistent time zone handling, which is a huge win, and also handles returning a full series with default values, which we also use.  Definitely not looking to move away from that for right now while changing the queries to point at new matviews across the board.